Further reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in audio code

cdumez opened the pull request:
#### 8b7ffa2d969a68ef098e98bc34ba1955b83c9bff
<pre>
Further reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in audio code
<a href="https://bugs.webkit.org/show_bug.cgi?id=285795">https://bugs.webkit.org/show_bug.cgi?id=285795</a>

Reviewed by Darin Adler.

* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
(PAL::CMBlockBufferGetDataSpan):
* Source/WebCore/platform/audio/DirectConvolver.cpp:
(WebCore::DirectConvolver::process):
* Source/WebCore/platform/audio/DownSampler.cpp:
(WebCore::DownSampler::process):
* Source/WebCore/platform/audio/SincResampler.cpp:
(WebCore::SincResampler::process):
(WebCore::SincResampler::convolve):
* Source/WebCore/platform/audio/SincResampler.h:
* Source/WebCore/platform/audio/UpSampler.cpp:
(WebCore::UpSampler::process):
* Source/WebCore/platform/audio/VectorMath.cpp:
(WebCore::VectorMath::dotProduct):
* Source/WebCore/platform/audio/VectorMath.h:
* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
(WebCore::passthroughInputDataCallback):
(WebCore::AudioFileReader::decodeWebMData const):
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.cpp:
(WebCore::WebAudioBufferList::setSampleCountWithBlockBuffer):
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::sharedBufferFromCMBlockBuffer):
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::decodeSampleInternal):

Canonical link: <a href="https://commits.webkit.org/288768@main">https://commits.webkit.org/288768@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4ce0295061830988b25aadae4ba81224dd0c3fbe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84348 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3973 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38653 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89427 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35357 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86433 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4058 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11950 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65600 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23437 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87394 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3048 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76637 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45894 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2999 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30867 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34406 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73889 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31636 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90808 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11615 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8441 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74050 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11842 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72461 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73250 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17579 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16023 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2990 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13062 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11567 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17043 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11416 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14892 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13189 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
```diff
@@ -409,6 +411,19 @@ SOFT_LINK_CONSTANT_MAY_FAIL_FOR_HEADER(PAL, CoreMedia, kCMFormatDescriptionExten
+    size_t lengthAtOffset = 0;
+    if (auto error = PAL::CMBlockBufferGetDataPointer(theBuffer, offset, &lengthAtOffset, nullptr, &data))
+        return { };
+    return byteCast<uint8_t>(unsafeMakeSpan(data, lengthAtOffset));
```
I'd be tempted to do the `byteCast` on the pointer rather than the `span` to reduce the chance of unnecessary copying.

cdumez committed:
Further reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in audio code
https://bugs.webkit.org/show_bug.cgi?id=285795

Reviewed by Darin Adler.

* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
(PAL::CMBlockBufferGetDataSpan):
* Source/WebCore/platform/audio/DirectConvolver.cpp:
(WebCore::DirectConvolver::process):
* Source/WebCore/platform/audio/DownSampler.cpp:
(WebCore::DownSampler::process):
* Source/WebCore/platform/audio/SincResampler.cpp:
(WebCore::SincResampler::process):
(WebCore::SincResampler::convolve):
* Source/WebCore/platform/audio/SincResampler.h:
* Source/WebCore/platform/audio/UpSampler.cpp:
(WebCore::UpSampler::process):
* Source/WebCore/platform/audio/VectorMath.cpp:
(WebCore::VectorMath::dotProduct):
* Source/WebCore/platform/audio/VectorMath.h:
* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
(WebCore::passthroughInputDataCallback):
(WebCore::AudioFileReader::decodeWebMData const):
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.cpp:
(WebCore::WebAudioBufferList::setSampleCountWithBlockBuffer):
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::sharedBufferFromCMBlockBuffer):
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::decodeSampleInternal):

Canonical link: https://commits.webkit.org/288768@main
